File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] result = new int[nums.length]; 
        // for(int i = 0; i < nums.length; i++){
        //     int num = 1;
        //     for(int j = 0; j < nums.length; j++){
        //         if(i == j){
        //             continue;
        //         }
        //         num *= nums[j];
        //     }
        //     result[i] = num;
        // }

        //sliding window
        int prev = nums[0];
        result[0] = 1;
        for(int i = 1; i < nums.length; i++){
            result[i] = prev;
            prev *= nums[i];
        }
        //1,2,3,4
        //     i
        int next= nums[nums.length-1]; 
        for(int i = nums.length-2; i >= 0; i--){
            result[i] *= next;
            next *= nums[i]; 
        }
        return result;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        left_products = [0] * n
        right_products = [0] * n
        final_answer = [0] * n

        temp = 1
        for i in range(n):
            left_products[i] = temp  
            temp *= nums[i]          

        temp = 1  
        for j in range(n - 1, -1, -1): 
            right_products[j] = temp 
            temp *= nums[j]         

        
        for i in range(n):
            final_answer[i] = left_products[i] * right_products[i] 

        return final_answer

            